tpm2/test/src/cipher/aes_keywrap_test.c

```diff
@@ -0,0 +1,539 @@
+                                               ct_data_len,
+                                               &out,
+                                               &out_len);
+            exp_result = pt_data;
+            exp_result_len = pt_data_len;
+          }
+
+          else
+          {
+            printf("ERROR: test vector set for file (%s) associated with ",
+                   aes_keywrap_vectors.sets[i].path);
+            printf("function to test(%s)\n",
+                   aes_keywrap_vectors.sets[i].func_to_test);
+            CU_FAIL("Test vector file linked to invalid function to test");
+            // don't get any more vectors from this file - can't apply them
+            done_with_test_vector_file = true;
+          }
+
+          // if vector successfully applied
+          if (rc != -1)
+          {
+            // consolidate results of applying test vector into a single assertion
+            if (result_bool == false)
+            {
+              // check if a test vector expected to fail, passed
+              if (rc == 0)
+              { 
+                vector_passed = false;
+                printf("expected fail, passed: %d\n", test_vector_count);
+              }
+            }
+            else
+            {
+              // check if a test vector expected to pass, failed
+              if (rc != 0)
+              {
+                vector_passed = false;
+                printf("expected pass, failed: %d\n", test_vector_count);
+              }
+
+              // check for unexpected size of decrypted result
+              if (out_len != exp_result_len)
+              {
+                vector_passed = false;
+                printf("unexpected size for result: %d\n", test_vector_count);
+              }
+
+              // check that expected result matches (byte for byte)
+              for (int j = 0; j < out_len; j++)
+              {
+                if (out[j] != exp_result[j])
+                {
+                  vector_passed = false;
+                  printf("expected result mismatch at byte %d: %d\n", j, test_vector_count);
+                }
+              }
+            }
+
+            CU_ASSERT(vector_passed);
+
+            // clean-up output_data byte array
+            if (rc == 0)
+            {
+              free(out);
+            }
+          }
+        }
+
+        else
+        {
+          // get_aes_gcm_test_vector_from_file() returned error - must be done
+          done_with_test_vector_file = true;
+        }
+      }
+
+      // Done with the test vector file (processed all vectors or reached max)
+      fclose(test_vector_fd[i]);
+
+      // Provide INFO: message indicating how many test vectors were applied
+      printf("INFO: %s - %d test vectors applied\n",
+              aes_keywrap_vectors.sets[i].path, test_vector_count);
+
+      // reset flag/counters for processing of next (if any) test vector file
+      done_with_test_vector_file = false;
```

Comment:
I think it would be cleaner to have these variables local to the `i` loop and set at the start of each iteration, rather than having to reset them at the end of each iteration.

Reply:
I moved the test_vector_count and done_with_test_vector_file variable declaration and initialization to just before the while loop used to process a test vector file ... much cleaner. In looking over this code, though, I noticed a number of other things that I thought might also improve it.

I also moved the check for (test_vector_count > MAX_KEYWRAP_TEST_VECTOR_COUNT) inside the while loop and set done_with_test_vector_file true, when exceeded. This simplifies the while condition to (!done_with_test_vector_file).

I moved the declaration and initialization of vector_passed to immediately precede the nested if statements that use it to created a boolean that indicates if all of the checks for that test vector passed, so that a single CU_ASSERT() can be used to report pass/fail status for that vector. Previously, it was declared and initialized much earlier in the while loop code block.

I also removed some printf() statements that I think were debug remnants from when I was originally writing and testing this code.

Finally, although I don't think it should ever happen, I added an else condition with a CU_FAIL() message if the test vector application results in a return code of -1. I believe that any of the kmyth cipher functions that these vectors are applied to should return only 0 if successful or 1 if an error is encountered. As we initialize rc to -1 and then assign it to the return value of the function call represented by our test vector, it should remain -1 only if the return code assignment does not actually happen or the called function returns -1 unexpectedly.